tests/device_selector/device_selector_aspect.cpp

```diff
@@ -0,0 +1,469 @@
+  }
+}
+
+/**
+ * Tests a whether a given selector conforms to a given list of required
+ * aspects and a list of denied aspects. */
+template <typename Selector>
+void test_selector(const Selector& selector,
+                   const std::vector<sycl::aspect>& accept_list,
+                   const std::vector<sycl::aspect>& deny_list) {
+  if (device_exists(accept_list, deny_list)) {
+    test_selector_accept(selector, accept_list);
+    test_selector_deny(selector, deny_list);
+  } else {
+    check_selector_exception(selector);
+  }
+}
+
+/**
+ * Checks aspect selector's constructors: using an accept list
+ * (and no deny list), using an accept list and a deny list, using accepted
+ * variadic function arguments, and using accepted variadic template
+ * parameters. */
+template <sycl::aspect... Aspects>
+void check_aspect_selector(const std::vector<sycl::aspect>& deny_list) {
+  const std::vector<sycl::aspect> accept_list{Aspects...};
+#ifndef SYCL_CTS_COMPILING_WITH_DPCPP
+  test_selector(sycl::aspect_selector(accept_list), accept_list);
+  if (!deny_list.empty()) {
+    test_selector(sycl::aspect_selector(accept_list, deny_list), accept_list,
+                  deny_list);
+  }
+  test_selector(sycl::aspect_selector(Aspects...), accept_list);
+  test_selector(sycl::aspect_selector<Aspects...>(), accept_list);
+#endif  // SYCL_CTS_COMPILING_WITH_DPCPP
+};
+
+/** Functor that checks a selector with one aspect, no denied aspects */
+template <typename AspectT>
+class check_for_single_aspect {
+  static constexpr sycl::aspect aspect = AspectT::value;
+
+ public:
+  void operator()(const std::string& aspect_name) {
+    INFO("for aspect " << aspect_name);
+    check_aspect_selector<aspect>({});
+  }
+};
+
+/** Functor that checks a selector with two aspects, no denied aspects */
+template <typename Aspect1T, typename Aspect2T>
+class check_for_two_aspects {
+  static constexpr sycl::aspect aspect1 = Aspect1T::value;
+  static constexpr sycl::aspect aspect2 = Aspect2T::value;
+
+ public:
+  void operator()(const std::string& aspect1_name,
+                  const std::string& aspect2_name) {
+    INFO("for aspects " << aspect1_name << " and " << aspect2_name);
+    check_aspect_selector<aspect1, aspect2>({});
+  }
+};
+
+/**
+ * Returns a randomly-sized list of random denied aspects that are not part of
+ * the requested aspects \p selected_aspect_list. */
+template <typename Rng>
+std::vector<sycl::aspect> generate_denied_list(
+    const std::vector<sycl::aspect>& aspect_list,
+    const std::vector<sycl::aspect>& selected_aspect_list, Rng& rng) {
+  std::vector<sycl::aspect> denied_list;
+  for (unsigned int i = 0; i < aspect_list.size(); i++) {
+    bool is_selected =
+        std::find(selected_aspect_list.begin(), selected_aspect_list.end(),
+                  aspect_list[i]) != selected_aspect_list.end();
+    if (is_selected) {
+      continue;
+    }
+    unsigned int p = rng() % aspect_list.size();
```

Comment:
Added comments here. I have increased the chance to `1 / # non-selected aspects`, meaning that the expected number of denied aspects will be one (since no selected aspect can be in the denied list). From looking at the tested aspects it looks okay, most of the time zero, one, or two denied aspects are selected.